### ISSUE TITLE
Adding `.reverse()` method to `IndexMap` & `IndexSet`

### ISSUE THREAD

**Issue by linclelinkpart5** [bluss/indexmap#127]
Looking at the impl of `OrderMapCore`, `.sort_by()` saves the hash index, sorts the underlying `entries` vector, and then restores the hash index. Would this mean that a `.reverse()` method could be added to `OrderedMapCore` (and thus `IndexMap`/`IndexSet`)?

**Comment by cuviper** [comment]
Reverse would be possible, yes. In fact, we can probably do the fixup more efficiently than sort since the change is predictable -- just sweep the hashed indices and flip each to `len - index`.

**Comment by linclelinkpart5** [comment]
> Reverse would be possible, yes. In fact, we can probably do the fixup more efficiently than sort since the change is predictable -- just sweep the hashed indices and flip each to `len - index`.

Just to understand correctly, this manipulation of the indices would replace the saving/restoring of the hash index that `.sort_by()` does normally, correct?

I created the implementation, but tests are failing right now, and I'm not exactly sure why!

**Comment by cuviper** [comment]
> Just to understand correctly, this manipulation of the indices would replace the saving/restoring of the hash index that `.sort_by()` does normally, correct?

Yes.

> I created the implementation, but tests are failing right now, and I'm not exactly sure why!

Can you share the code and details on what's failing?

**Comment by linclelinkpart5** [comment]
Hey @cuviper, thanks for the info! I realized the cause of my test failures: I forgot that when inserting multiple key-value pairs into an `IndexSet`, any duplicated keys stay put, and only their values get updated. I was wrongly assuming that adding a duplicated key brought that KV pair to the right, order-wise.

I've created a PR, but it would still need the improved re-indexing you suggested.
